Readme.md
# Bytes utility

[![NPM Version][npm-image]][npm-url]
[![NPM Downloads][downloads-image]][downloads-url]
[![Build Status][travis-image]][travis-url]
<!--
[![Test Coverage][coveralls-image]][coveralls-url]
 -->
Utility to parse a size string in bytes (e.g. `'1kB'`, `'2KiB'`) to numeric (`1000`, `2048`) and vice-versa.

This is a fork of the [bytes][bytes-url] module, except it

 * supports all of both the binary and decimal prefixes defined by [ISO/IEC 80000-13:2008][binary-wiki]
 * supports [JEDEC][jedec-wiki], a legacy mode in which metric units have binary values
 * uses decimal metric units by default, which can be overridden per call or by default

TypeScript definitions included.

## Supported Units

Supported units are as follows and are case-insensitive. Note that only the abbreviation will be parsed/formatted, the full names are for the reader's understanding only.

### Metric

Also referred to as SI. See [Compatibility Binary](#compatibility-binary) for legacy definitions.

|      Value       | Abbr |   Name    |
| ---------------- | ---- | --------- |
| 1                | B    | byte      |
| 1000<sup>1</sup> | kB   | kilobyte  |
| 1000<sup>2</sup> | MB   | megabyte  |
| 1000<sup>3</sup> | GB   | gigabyte  |
| 1000<sup>4</sup> | TB   | terabyte  |
| 1000<sup>5</sup> | PB   | petabyte  |
| 1000<sup>6</sup> | EB   | exabyte   |
| 1000<sup>7</sup> | ZB   | zettabyte |
| 1000<sup>8</sup> | YB   | yottabyte |

[More info][metric-wiki]

### Binary

|      Value       | Abbr |   Name    |
| ---------------- | ---- | --------- |
| 1                | B    | byte      |
| 1024<sup>1</sup> | KiB  | kibibyte  |
| 1024<sup>2</sup> | MiB  | mebibyte  |
| 1024<sup>3</sup> | GiB  | gibibyte  |
| 1024<sup>4</sup> | TiB  | tebibyte  |
| 1024<sup>5</sup> | PiB  | pebibyte  |
| 1024<sup>6</sup> | EiB  | exbibyte  |
| 1024<sup>7</sup> | ZiB  | zebibite  |
| 1024<sup>8</sup> | YiB  | yobibite  |

[More info][binary-wiki]

### Compatibility Binary

Also referred to as JEDEC or legacy units.

Overwrites the lower units of the metric system with the commonly misused values, i.e. metric units will be binary instead of decimal.
This is the behavior of e.g. the Windows OS and [bytes][bytes-url].
Units greater than terabyte are not supported.

|      Value       | Abbr |   Name    |
| ---------------- | ---- | --------- |
| 1024<sup>1</sup> | kB   | kilobyte  |
| 1024<sup>2</sup> | MB   | megabyte  |
| 1024<sup>3</sup> | GB   | gigabyte  |
| 1024<sup>4</sup> | TB   | terabyte  |

[More info][jedec-wiki]

## Installation

This is a [Node.js](https://nodejs.org/en/) module available through the
[npm registry](https://www.npmjs.com/). Installation is done using the
[`npm install` command](https://docs.npmjs.com/getting-started/installing-npm-packages-locally):

```bash
npm install bytes-iec
```

## Usage

```js
var bytes = require('bytes-iec');
```

#### Modes

Passing a unit type as `mode` parameter in API calls determines

 * the set of units that will be favored by autodetection when no unit is specified
 * the value/size of metric units: Compatibility mode makes them base-2 instead of base-10

| Unit type                                     | `mode`                         |
| --------------------------------------------- | ------------------------------ |
| [Metric](#metric)                             | `'metric'` or `'decimal'`      |
| [Binary](#binary)                             | `'binary'`                     |
| [Compatibility Binary](#compatibility-binary) | `'compatibility'` or `'jedec'` |

#### bytes.format(number value, [options]): string|null

Format the given value in bytes into a string. If the value is negative, it's kept as such. If it's a float, it's rounded.

**Arguments**

| Name    | Type     | Description        |
|---------|----------|--------------------|
| value   | `number` | Value in bytes     |
| options | `Object` | Conversion options |

**Options**

| Property           | Type              | Description                                                                                   | Default           |
| ------------------ | ----------------- | --------------------------------------------------------------------------------------------- | ----------------- |
| decimalPlaces      | `number`｜`null`  | Maximum number of decimal places to include in output                                         | `2`               |
| fixedDecimals      | `boolean`｜`null` | Whether to always display the maximum number of decimal places, i.e. preserve trailing zeroes | `false`           |
| thousandsSeparator | `string`｜`null`  | What to separate large numbers with, e.g. `','`, `'.'`, `' '`, ...                            | `''`              |
| unit               | `string`｜`null`  | The unit in which the result will be returned: `'B'`, `'kB'`, `'KiB'`, ...                    | `''` (autodetect) |
| unitSeparator      | `string`｜`null`  | Separator between numeric value and unit                                                      | `''`              |
| mode               | `string`｜`null`  | Which mode to use (see [Modes](#modes))                                                       | `'metric'`        |

**Returns**

| Name    | Type             | Description                                     |
|---------|------------------|-------------------------------------------------|
| results | `string`｜`null` | Returns null upon error, string value otherwise. |

**Example**

```js
bytes(1000);
// output: '1kB'

bytes(1000, {thousandsSeparator: ' '});
// output: '1 000B'

bytes(1024);
// output: '1.02kB'

bytes(1024 * 1.7, {decimalPlaces: 0});
// output: '2KB'

bytes(1000, {unitSeparator: ' '});
// output: '1 kB'

bytes(2048, {mode: 'binary'});
// output: '2 KiB'

bytes(1024 * 1024 * 2, {unit: 'KiB'});
// output: '2048 KiB'

bytes(1024 * 1024 * 2, {unit: 'KB'});
// output: '2097.152 KB'

bytes(1024 * 1024 * 2, {unit: 'KB', mode: 'compatibility'});
// output: '2048 KB'
```

#### bytes.parse(string｜number value): number｜null

Parse the string value into an integer in bytes. If no unit is given, or `value`
is a number, it is assumed the value is in bytes.

If the value given has partial bytes, it's truncated (rounded down).

**Arguments**

| Name          | Type   | Description        |
|---------------|--------|--------------------|
| value   | `string`｜`number` | String to parse, or number in bytes |
| options | `Object` | Conversion options |

|       Property       |          Type         | Description | Default |
| -------------------- | --------------------- | ----------- |---------|
| mode   | `string`｜`null` | Which mode to use (see [Modes](#modes)) | `'metric'` |

**Returns**

| Name    | Type        | Description             |
|---------|-------------|-------------------------|
| results | `number`｜`null` | Returns null upon error, value in bytes otherwise. |

**Example**

```js
bytes('1kB');
// output: 1024

bytes('1024');
// output: 1024

bytes('1.0001 kB');
// output: 1000
bytes('1.0001 KiB');
// output: 1024

bytes('1kB', {mode: 'jedec'});
// output: 1024
```

#### bytes.withDefaultMode(string mode): object

Returns a new copy of the `bytes-iec` module, but with the given mode as the default.

**Arguments**

| Name          | Type     | Description        |
|---------------|----------|--------------------|
| mode          | `string` | Default mode to use (see [Modes](#modes))   |

**Returns**

| Name    | Type        | Description             |
|---------|-------------|-------------------------|
| results | `object` | Returns the byte.js module, with a default mode. |

**Example**

```js
var bytes = require('bytes').withDefaultMode('jedec');

bytes('1kB');
// output: 1024

bytes('1KiB');
// output: 1024

bytes(1024);
// output: 1 kB

bytes(1024, {mode: 'metric'});
// output: 1.02kB

bytes('1kB', {mode: 'metric'});
// output: 1000
```

## License

[MIT](LICENSE)

<!--
[coveralls-image]: https://badgen.net/coveralls/c/github/visionmedia/bytes.js/master
[coveralls-url]: https://coveralls.io/r/visionmedia/bytes.js?branch=master
-->

[downloads-image]: https://badgen.net/npm/dm/bytes-iec
[downloads-url]: https://npmjs.org/package/bytes-iec
[npm-image]: https://badgen.net/npm/node/bytes-iec
[npm-url]: https://npmjs.org/package/bytes-iec
[travis-image]: https://badgen.net/travis/saevon/bytes.js/master
[travis-url]: https://travis-ci.org/saevon/bytes.js
[bytes-url]: https://github.com/visionmedia/bytes.js
[binary-wiki]: https://en.wikipedia.org/wiki/Binary_prefix
[metric-wiki]: https://en.wikipedia.org/wiki/Metric_prefix
[jedec-wiki]: https://en.wikipedia.org/wiki/JEDEC_memory_standards#Unit_prefixes_for_semiconductor_storage_capacity


History.md
Bytes-IEC
=================
* Adds ISO/IEC 80000-13:2008 format
- Adds compatibility mode for parsing older programs

3.1.1 / Pending
==================

  * Adds jedec as an alias for compatibility mode
  * Improves error checking for modes
  * Adds TypeScript definitions

3.1.0 / 2019-01-22
==================

  * Add petabyte (`pb`) support

3.0.0 / 2017-08-31
==================

  * Change "kB" to "KB" in format output
  * Remove support for Node.js 0.6
  * Remove support for ComponentJS

2.5.0 / 2017-03-24
==================

  * Add option "unit"

2.4.0 / 2016-06-01
==================

  * Add option "unitSeparator"

2.3.0 / 2016-02-15
==================

  * Drop partial bytes on all parsed units
  * Fix non-finite numbers to `.format` to return `null`
  * Fix parsing byte string that looks like hex
  * perf: hoist regular expressions

2.2.0 / 2015-11-13
==================

  * add option "decimalPlaces"
  * add option "fixedDecimals"

2.1.0 / 2015-05-21
==================

  * add `.format` export
  * add `.parse` export

2.0.2 / 2015-05-20
==================

  * remove map recreation
  * remove unnecessary object construction

2.0.1 / 2015-05-07
==================

  * fix browserify require
  * remove node.extend dependency

2.0.0 / 2015-04-12
==================

  * add option "case"
  * add option "thousandsSeparator"
  * return "null" on invalid parse input
  * support proper round-trip: bytes(bytes(num)) === num
  * units no longer case sensitive when parsing

1.0.0 / 2014-05-05
==================

 * add negative support. fixes #6

0.3.0 / 2014-03-19
==================

 * added terabyte support

0.2.1 / 2013-04-01
==================

  * add .component

0.2.0 / 2012-10-28
==================

  * bytes(200).should.eql('200b')

0.1.0 / 2012-07-04
==================

  * add bytes to string conversion [yields]


test/bytes.js
'use strict';

var assert = require('assert');
var bytes = require('../index.js');

describe('Test constructor', function(){
  it('Expect a function', function(){
    assert.equal(typeof bytes, 'function');
  });

  it('Should return null if input is invalid', function(){
    assert.strictEqual(bytes(undefined), null);
    assert.strictEqual(bytes(null), null);
    assert.strictEqual(bytes(true), null);
    assert.strictEqual(bytes(false), null);
    assert.strictEqual(bytes(NaN), null);
    assert.strictEqual(bytes(function(){}), null);
    assert.strictEqual(bytes({}), null);
  });
  it('Shoud be able to parse a string into a number', function(){
    // This function is testes more accurately in another test suite
    assert.equal(bytes('1kB'), 1000);
  });

  it('Should convert a number into a string', function(){
    // This function is testes more accurately in another test suite
    assert.equal(bytes(1000), '1kB');
  });

  it('Should convert a number into a string with options', function(){
    // This function is tested more accurately in another test suite
    assert.equal(bytes(1000, {thousandsSeparator: ' ', unit: "b"}), '1 000B');
    assert.equal(bytes(1000, {unitSeparator: '---'}), '1---kB');
  });
});

describe('Test withDefaultMode', function(){
  var jedec = bytes.withDefaultMode('jedec');
  var binary = bytes.withDefaultMode('binary');
  var metric = bytes.withDefaultMode('metric');

  it('Expect a function', function(){
    assert.equal(typeof jedec, 'function');
  });

  it('Should still be able to parse a string into a number', function(){
    assert.equal(jedec('1kB'), 1024);
  });

  it('Should still accept other modes', function(){
    assert.equal(jedec('1kB', {mode: 'metric'}), 1000);
  });

  it('Should still convert a number into a string', function(){
    assert.equal(jedec(1024), '1kB');
  });

  it('Should run parse properly', function(){
    assert.equal(jedec.parse('1kB'), 1024);
  });

  it('Should run format properly', function(){
    assert.equal(jedec.format(1024), '1kB');
  });

  it('Should not affect older modules', function(){
    assert.equal(bytes.format(1024), '1.02kB');
    assert.equal(jedec.format(1024), '1kB');
    assert.equal(binary.format(1024), '1KiB');
    assert.equal(metric.format(1000), '1kB');
  });

  it('Allow aliases: compatibility', function(){
    var compatibility = bytes.withDefaultMode('compatibility');
    assert.equal(compatibility.format(1024), '1kB');
    assert.equal(compatibility.parse('1kB'), 1024);

    assert.equal(metric.parse('1kB', {mode: 'compatibility'}), 1024);
    assert.equal(metric.format(1024, {mode: 'compatibility'}), '1kB');
  });

  it('Allow aliases: decimal', function(){
    var decimal = bytes.withDefaultMode('decimal');
    assert.equal(decimal.parse('1KB'), 1000);
    assert.equal(decimal.format(1000), '1kB');

    assert.equal(binary.parse('1KB', {mode: 'decimal'}), 1000);
    assert.equal(binary.format(1000, {mode: 'decimal'}), '1kB');
  });

});



describe('Test bad modes', function(){
  it('should fail with unknown modes', function(){
    assert.throws(function() {
      bytes.withDefaultMode('unknown');
    }, Error, "bytes.js: invalid mode passed in: unknown");
    assert.throws(function() {
      bytes.parse('1kb', {mode: 'unknown'});
    }, Error, "bytes.js: invalid mode passed in: unknown");
    assert.throws(function()  {
      bytes.format(1, {mode: 'unknown'});
    }, Error, "bytes.js: invalid mode passed in: unknown");
  })
})






test/byte-parse.js
'use strict';

var assert = require('assert');
var bytes = require('..');

describe('Test byte parse function', function(){
  it('Should return null if input is invalid', function(){
    assert.strictEqual(bytes.parse(undefined), null);
    assert.strictEqual(bytes.parse(null), null);
    assert.strictEqual(bytes.parse(true), null);
    assert.strictEqual(bytes.parse(false), null);
    assert.strictEqual(bytes.parse(NaN), null);
    assert.strictEqual(bytes.parse(function(){}), null);
    assert.strictEqual(bytes.parse({}), null);
    assert.strictEqual(bytes.parse('0x11'), null);
  });

  it('Should not accept partial units', function(){
    assert.strictEqual(bytes.parse('10Mi'), null);
  });

  it('Should parse raw number', function(){
    assert.strictEqual(bytes.parse(0), 0);
    assert.strictEqual(bytes.parse(-1), -1);
    assert.strictEqual(bytes.parse(1), 1);
    assert.strictEqual(bytes.parse(10.5), 10.5);
  });

  it('Should parse bytes', function(){
    assert.equal(bytes.parse('1B'), 1);
    assert.equal(bytes.parse('123B'), 123);
  });

  it('Should parse KiB', function(){
    assert.equal(bytes.parse('1KiB'), 1 * Math.pow(1024, 1));
    assert.equal(bytes.parse('0.5KiB'), 0.5 * Math.pow(1024, 1));
    assert.equal(bytes.parse('1.5KiB'), 1.5 * Math.pow(1024, 1));
  });

  it('Should parse upper binary units', function(){
    assert.equal(bytes.parse('1MiB'), 1 * Math.pow(1024, 2));
    assert.equal(bytes.parse('1GiB'), 1 * Math.pow(1024, 3));
    assert.equal(bytes.parse('1TiB'), 1 * Math.pow(1024, 4));
    assert.equal(bytes.parse('1PiB'), 1 * Math.pow(1024, 5));
    assert.equal(bytes.parse('1EiB'), 1 * Math.pow(1024, 6));
    assert.equal(bytes.parse('1ZiB'), 1 * Math.pow(1024, 7));
    assert.equal(bytes.parse('1YiB'), 1 * Math.pow(1024, 8));
  });

  it('Should parse kB', function(){
    assert.equal(bytes.parse('1kB'), 1 * Math.pow(1000, 1));
    assert.equal(bytes.parse('0.5kB'), 0.5 * Math.pow(1000, 1));
    assert.equal(bytes.parse('1.5kB'), 1.5 * Math.pow(1000, 1));
  });

  it('Should parse upper metric units', function(){
    assert.equal(bytes.parse('1MB'), 1 * Math.pow(1000, 2));
    assert.equal(bytes.parse('1GB'), 1 * Math.pow(1000, 3));
    assert.equal(bytes.parse('1TB'), 1 * Math.pow(1000, 4));
    assert.equal(bytes.parse('1PB'), 1 * Math.pow(1000, 5));
    assert.equal(bytes.parse('1EB'), 1 * Math.pow(1000, 6));
    assert.equal(bytes.parse('1ZB'), 1 * Math.pow(1000, 7));
    assert.equal(bytes.parse('1YB'), 1 * Math.pow(1000, 8));
  });

  it('Should parse jedec units', function(){
    assert.equal(bytes.parse('1KB', {mode: 'jedec'}), 1 * Math.pow(1024, 1));
    assert.equal(bytes.parse('1MB', {mode: 'jedec'}), 1 * Math.pow(1024, 2));
    assert.equal(bytes.parse('1GB', {mode: 'jedec'}), 1 * Math.pow(1024, 3));
    assert.equal(bytes.parse('1TB', {mode: 'jedec'}), 1 * Math.pow(1024, 4));
  });

  it('Should parse remaining normal units in jedec mode', function(){
    assert.equal(bytes.parse('1B', {mode: 'jedec'}), 1);
    assert.equal(bytes.parse('1YB', {mode: 'jedec'}), 1 * Math.pow(1000, 8));
  });

  it('Should assume bytes when no units', function(){
    assert.equal(bytes.parse('0'), 0);
    assert.equal(bytes.parse('-1'), -1);
    assert.equal(bytes.parse('1024'), 1024);
  });

  it('Should accept negative values', function(){
    assert.equal(bytes.parse('-1'), -1);
    assert.equal(bytes.parse('-1024'), -1024);
    assert.equal(bytes.parse('-1.5TB'), -1.5 * Math.pow(1000, 4));
  });

  it('Should drop partial bytes', function(){
    assert.equal(bytes.parse('1.1b'), 1);
    assert.equal(bytes.parse('1.0001kb'), 1000);
  });

  it('Should allow whitespace', function(){
    assert.equal(bytes.parse('1 TiB'), 1 * Math.pow(1024, 4));
  });

  it('Should parse case-insensitively', function(){
    assert.equal(bytes.parse('1kib'), 1 * Math.pow(1024, 1));
    assert.equal(bytes.parse('1KiB'), 1 * Math.pow(1024, 1));
    assert.equal(bytes.parse('1Kib'), 1 * Math.pow(1024, 1));
    assert.equal(bytes.parse('1kiB'), 1 * Math.pow(1024, 1));
    assert.equal(bytes.parse('1KIB'), 1 * Math.pow(1024, 1));
  });
});


test/types.spec.ts
import assert from "assert";
import bytes from "bytes-iec";

/*
this test is intended more to verify type definitions than functionality
*/

describe("Entrypoints", () => {
  it("is callable (`format`)", () => {
    assert.strictEqual(typeof bytes, "function");
  });

  it("has `format`, `parse`, and `withDefaultMode`", () => {
    assert.strictEqual(typeof bytes.format, "function");
    assert.strictEqual(typeof bytes.parse, "function");
    assert.strictEqual(typeof bytes.withDefaultMode, "function");
  });

  it("accepts all expected options to `format`", () => {
    const options: Required<Parameters<typeof bytes.format>[1]> = {
      decimalPlaces: 2,
      fixedDecimals: true,
      mode: "binary",
      thousandsSeparator: "",
      unit: "kiB",
      unitSeparator: ""
    };
    const formatted = bytes.format(1024, options);

    assert.strictEqual(formatted, "1.00KiB");
  });

  it("accepts all expected options to `parse`", () => {
    const options: Required<Parameters<typeof bytes.parse>[1]> = {
      mode: "binary"
    };
    const parsed = bytes.parse("1KiB", options);

    assert.strictEqual(parsed, 1024);
  });

  it("accepts all expected options to `withDefaultMode`", () => {
    const parsed = bytes.withDefaultMode("compatibility").parse("1KB");

    assert.strictEqual(parsed, 1024);
  });
});


index.js
/*!
 * bytes
 * Copyright(c) 2012-2014 TJ Holowaychuk
 * Copyright(c) 2015 Jed Watson
 * MIT Licensed
 */

'use strict';

/**
 * Module exports.
 * @public
 */
module.exports = withDefaultMode('compatibility');

/**
 * Module variables.
 * @private
 */

var formatThousandsRegExp = /\B(?=(\d{3})+(?!\d))/g;

var formatDecimalsRegExp = /(?:\.0*|(\.[^0]+)0+)$/;

var allUnits = {
  // Metric units
  b: {val: 1, longName: 'byte', name: 'B'},
  kb: {val: Math.pow(1000, 1), longName: 'kilobyte', name: "kB"},
  mb: {val: Math.pow(1000, 2), longName: 'megabyte', name: "MB"},
  gb: {val: Math.pow(1000, 3), longName: 'gigabyte', name: "GB"},
  tb: {val: Math.pow(1000, 4), longName: 'terabyte', name: "TB"},
  pb: {val: Math.pow(1000, 5), longName: 'petabyte', name: "PB"},
  eb: {val: Math.pow(1000, 6), longName: 'exabyte', name: "EB"},
  zb: {val: Math.pow(1000, 7), longName: 'zettabyte', name: "ZB"},
  yb: {val: Math.pow(1000, 8), longName: 'yottabyte', name: "YB"},

  // Binary units
  kib: {val: Math.pow(1024, 1), longName: 'kibibyte', name: "KiB"},
  mib: {val: Math.pow(1024, 2), longName: 'mebibyte', name: "MiB"},
  gib: {val: Math.pow(1024, 3), longName: 'gibibyte', name: "GiB"},
  tib: {val: Math.pow(1024, 4), longName: 'tebibyte', name: "TiB"},
  pib: {val: Math.pow(1024, 5), longName: 'pebibyte', name: "PiB"},
  eib: {val: Math.pow(1024, 6), longName: 'exbibyte', name: "EiB"},
  zib: {val: Math.pow(1024, 7), longName: 'zebibyte', name: "ZiB"},
  yib: {val: Math.pow(1024, 8), longName: 'yobibyte', name: "YiB"},

  // Compatibility units
  _kb: {val: Math.pow(1024, 1), longName: 'kilobyte', name: "kB"},
  _mb: {val: Math.pow(1024, 2), longName: 'megabyte', name: "MB"},
  _gb: {val: Math.pow(1024, 3), longName: 'gigabyte', name: "GB"},
  _tb: {val: Math.pow(1024, 4), longName: 'terabyte', name: "TB"},
};


function _invertUnitMap(expectedUnits) {
  var orderedUnits = [];

  for (var i=0; i < expectedUnits.length; i++) {
    var unitData = expectedUnits[i];

    orderedUnits.push(allUnits[unitData]);
  }

  orderedUnits.sort(function sortLargest(a, b) {
    return (b.val - a.val);
  });
  return orderedUnits;
}

var metricMap = _invertUnitMap([
  'b',
  'kb',
  'mb',
  'gb',
  'tb',
  'pb',
  'eb',
  'zb',
  'yb',
]);
var binaryMap = _invertUnitMap([
  'b',
  'kib',
  'mib',
  'gib',
  'tib',
  'pib',
  'eib',
  'zib',
  'yib',
]);
var compatibilityMap = _invertUnitMap([
  'b',
  '_kb',
  '_mb',
  '_gb',
  '_tb',
]);


var parseRegExp = /^((-|\+)?(\d+(?:\.\d+)?)) *((k|m|g|t|p|e|z|y)?i?(b))?$/i;

/**
 * Convert the given value in bytes into a string or parse to string to an integer in bytes.
 *
 * @param {string|number} value
 * @param {{
 *  case: [string],
 *  decimalPlaces: [number]
 *  fixedDecimals: [boolean]
 *  thousandsSeparator: [string]
 *  unitSeparator: [string]
 *  mode: [string]
 *  unit: [string]
 *  }} [options] bytes options.
 *
 * @returns {string|number|null}
 */

function bytes(value, options) {
  if (typeof value === 'string') {
    return parse(value, options);
  }

  if (typeof value === 'number') {
    return format(value, options);
  }

  return null;
}

/**
 * Creates a library wrapper that forces a certain mode to be default
 *
 * @param {string} mode
 */
function withDefaultMode(mode) {
  // Adds the default mode to the method
  function setDefault(options) {
    options = (options !== undefined ? options : {});

    if (options.mode === undefined) {
      options.mode = mode;
    }

    return options;
  }

  function bytesWithDefault(value, options) {
    return bytes(value, setDefault(options));
  }

  bytesWithDefault.format = function formatWithDefault(value, options) {
    return format(value, setDefault(options));
  };

  bytesWithDefault.parse = function formatWithDefault(value, options) {
    return parse(value, setDefault(options));
  };

  // Also allow it to act completely like the original module
  // So we can still change the default mode (but only by makind a new module)
  bytesWithDefault.withDefaultMode = withDefaultMode;

  return bytesWithDefault;
}

/**
 * Format the given value in bytes into a string.
 *
 * If the value is negative, it is kept as such. If it is a float,
 * it is rounded.
 *
 * @param {number} value
 * @param {object} [options]
 * @param {number} [options.decimalPlaces=2]
 * @param {number} [options.fixedDecimals=false]
 * @param {string} [options.thousandsSeparator=]
 * @param {string} [options.unit=]
 * @param {string} [options.unitSeparator=]
 * @param {string} [options.mode='metric']
 * @param {string} [options.unit=]
 *
 * @returns {string|null}
 * @public
 */

function format(value, options) {
  if (!Number.isFinite(value)) {
    return null;
  }

  var mag = Math.abs(value);
  var thousandsSeparator = (options && options.thousandsSeparator) || '';
  var unitSeparator = (options && options.unitSeparator) || '';
  var decimalPlaces = (options && options.decimalPlaces !== undefined) ? options.decimalPlaces : 2;
  var fixedDecimals = Boolean(options && options.fixedDecimals);
  var mode = (options && options.mode) || 'metric';
  var expectedUnit = (options && options.unit);

  // Find which set of units we're converting to
  var unitMap = _getUnitMap(mode);

  // Find what to convert to
  var conversionUnit;
  if (expectedUnit !== undefined) {
    // The user specified an expected unit
    conversionUnit = _findUnit(allUnits, expectedUnit, mode);
  } else {
    // The user wants the most appropriate unit
    conversionUnit = _findLargestUnit(unitMap, mag);
  }

  // Convert the unit
  var val = value / conversionUnit.val;
  var str = val.toFixed(decimalPlaces);
  var unit = conversionUnit.name;

  if (!fixedDecimals) {
    str = str.replace(formatDecimalsRegExp, '$1');
  }

  if (thousandsSeparator) {
    str = str.replace(formatThousandsRegExp, thousandsSeparator);
  }

  return str + unitSeparator + unit;
}

/**
 * Returns an array that tells you what units to convert to and in what order.
 */
function _getUnitMap(mode) {
  // Allow the mode synonym
  if (mode == 'decimal' || mode == 'metric') {
    return metricMap;
  } else if (mode == 'compatibility') {
    return compatibilityMap;
  } else if (mode == 'binary') {
    return binaryMap;
  } else {
    throw "bytes.js: invalid mode passed in: " + mode;
  }
}

/**
 * Finds the given unit in the unit map
 */
function _findUnit(unitMap, unit, mode) {
  var conversionUnit;

  unit = unit.toLowerCase();

  // Check the compatibility units first
  if (mode == 'compatibility') {
    conversionUnit = allUnits['_' + unit];
  }

  // Otherwise check the rest of the units
  if (conversionUnit === undefined) {
    conversionUnit = allUnits[unit];
  }

  // Finally if we still didn't find the unit, its an error
  if (conversionUnit === undefined) {
    throw "byte.js: unit not found: " + unit;
  }

  return conversionUnit;
}

/**
 * Finds the largest unit that the value can be converted to
 *
 * @param {Array} unitMap Array of units sorted with the largest units first
 * @param {int}   value
 *
 * @returns {}
 * @private
 */
function _findLargestUnit(unitMap, value) {
  for (var i=0; i < unitMap.length; i++) {
    var unitData = unitMap[i];

    // Find the largest unit to convert to, otherwise use the last one
    if (value >= unitData.val || i >= unitMap.length - 1) {
      return unitData;
    }
  }

}

/**
 * Parse the string value into an integer in bytes.
 *
 * If no unit is given, it is assumed the value is in bytes.
 *
 * @param {number|string} val
 * @param {object} [options]
 * @param {string} [options.mode='metric']
 *
 * @returns {number|null}
 * @public
 */
function parse(val, options) {
  if (typeof val === 'number' && !isNaN(val)) {
    return val;
  }

  if (typeof val !== 'string') {
    return null;
  }

  var useCompatibility = (options && options.mode == 'compatibility') ? true : false;

  // Test if the string passed is valid
  var results = parseRegExp.exec(val);
  var floatValue;
  var unit;

  // No number was extracted from the input string
  if (!results) {
    return null;
  }

  // Retrieve the value and the unit
  floatValue = parseFloat(results[1]);
  unit = results[4];

  // Default unit if none are specified
  if (unit === undefined) {
    unit = 'b';
  }

  // Make sure we're case-insensitive
  unit = unit.toLowerCase();

  var unitData;
  // If we're using compatibility units, try those first
  if (useCompatibility) {
    unitData = allUnits['_' + unit];
  }

  // Try to get the unit from the normal units
  if (unitData === undefined) {
    unitData = allUnits[unit];
  }

  return Math.floor(unitData.val * floatValue);
}
